src/main/kotlin/com/avail/optimizer/values/L2SemanticPrimitiveInvocation.kt

```diff
@@ -0,0 +1,145 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.avail.optimizer.values
+
+import com.avail.descriptor.representation.AvailObject
+import com.avail.interpreter.Primitive
+import com.avail.utility.Casts
+
+/**
+ * An [L2SemanticValue] which represents the result produced by a [Primitive]
+ * when supplied a list of argument [L2SemanticValue]s.  The primitive must be
+ * stable (same result), pure (no side-effects), and successful for the supplied
+ * arguments.
+ *
+ * @author Mark van Gulik &lt;mark@availlang.org&gt;
+ * @author Todd L Smith &lt;todd@availlang.org&gt;
+ *
+ * @constructor
+ * Create a new `L2SemanticPrimitiveInvocation` semantic value.
+ *
+ * @property primitive
+ *   The [Primitive] whose invocation is being represented.
+ * @property argumentSemanticValues
+ *   The [List] of [L2SemanticValue]s that represent the arguments to the
+ *   invocation of the primitive.
+ * @param primitive
+ *   The primitive whose invocation is being represented.
+ * @param argumentSemanticValues
+ *   The semantic values supplied as arguments.
+ */
+class L2SemanticPrimitiveInvocation internal constructor(
+	@JvmField val primitive: Primitive,
+	@JvmField val argumentSemanticValues: List<L2SemanticValue>)
+		: L2SemanticValue(computeHash(primitive, argumentSemanticValues))
+{
+	init
+	{
+		assert(primitive.hasFlag(Primitive.Flag.CanFold))
+	}
+
+	override fun equals(obj: Any?): Boolean
+	{
+		if (obj !is L2SemanticPrimitiveInvocation)
+		{
+			return false
+		}
+		val invocation =
+			Casts.cast<Any, L2SemanticPrimitiveInvocation>(obj)
+		return (primitive === invocation.primitive
+			&& argumentSemanticValues == invocation.argumentSemanticValues)
+	}
+
+	override fun toString(): String = buildString {
+		append(primitive.fieldName())
+		append('(')
+		var first = true
+		for (arg in argumentSemanticValues)
+		{
+			if (!first)
+			{
+				append(", ")
+			}
+			append(arg)
+			first = false
+		}
+		append(')')
+	}
+
+	override fun transform(
+		semanticValueTransformer: Function1<L2SemanticValue, L2SemanticValue>,
+		frameTransformer: Function1<Frame, Frame>): L2SemanticValue
+	{
+		val numArgs = argumentSemanticValues.size
+		val newArguments =
+			argumentSemanticValues.map {
+				it.transform(semanticValueTransformer, frameTransformer)
+			}.toMutableList()
+
+		for (i in 0 until numArgs)
```

Comment:
```
if ((0 until numArgs).all { newArgumenst[i] == argumentSemanticValues[i] }) {
    return this
}
return L2SemanticPrimitiveInvocation(primitive, newArguments)
```

or possibly try something with infix zip?

Reply:
wasn't really sure how to work zip into this. Maybe it is because it is kinda late, so I just took your suggestion.